Docs: OAuth2 sequence with Edge Proxy Server (#8)
fluid-project/personal-data-server: doc/StaticWorkflow.md

# Preferences Edge Proxy Workflow

> A workflow that the Personal Data Storage or Preferences Servier supports

Reviewer: ```suggestion
A workflow that the Personal Data Storage or Preferences Server supports
```

> A workflow that the Personal Data Storage or Preferences Servier supports
> i where a static site makes save/retrieve requests for the user's preferences.

Reviewer: ```suggestion
A workflow that the Personal Data Storage or Preferences Server supports
is where a static site makes save/retrieve requests for the user's preferences.
```

> An example is where a user changes their UIO options and wants to save them.

Reviewer: ```suggestion
An example is where a user changes their UIO preferences and wants to save them.
```

Included in this workflow is an OAuth2 authorization sequence where
users are authenticated by a third party single sign on (SSO) provider, such as

> Google or Github.

Reviewer: ```suggestion
Google or GitHub.
```

Author: Thanks for catching all of my typos, and for the suggestions.  I have modified accordingly.

      valid by checking that it matches the value it sent to Google in step 2i.
   6. [Exchange Authorization Code](https://developers.google.com/identity/protocols/oauth2/openid-connect#exchangecode):
      The Preferences Server requests exchanging the Authorization Code for an
      Access Token and ID Token. This includes the following:
      * `code`, the Authorization Code sent from the previous response from
         Google
      * `client_id` for the Preferences Server (same value as steps 2i)
      * `redirect_uri` which is the endpoint on the Preferences Server that
         will receive the response from Google (same value as step 2i)
      * `grant_type` which must be set to `authorization_code`
   7. Google responds at the previously specified redirect uri with the Access
      Token and ID Token
      * The `access_token` can be used by the Preferences Server to access the Google API
      * The `id_token` is a [JWT](https://tools.ietf.org/html/rfc7519)
        identifying the Resource Owner and is signed by Google.
3. Authenticate and Authorize UIO
   1. Using the `id_token` the Preferences Server authenticates the Resource

>       Owner

Reviewer: We talked about this recently, but adding here for reference. For services such as GitHub that do not support OpenID Connect, there will be no `id_token` returned. It's handy to use the `id_token` if available, but we can use the `access_token` if not. However, we may need to make additional requests to the SSO to verify it's still valid.

Author: Yes, we discussed this at today's meeting, and decided not to use either.  The Preferences Server will generate a session key after it receives the access token from the SSO provider.  This is more secure since it keeps the values separate -- the access token is a "secret" shared between the Preferences Server and Google.  The Preferences Server will persist this session key, and pass it back to the Edge Proxy and UIO.  They, in turn, will use it to make requests of the Preferences Server.  I've added the changes to the document that explain this approach.

> However, we may need to make additional requests to the SSO to verify it's still valid.

We _will_ have to make additional requests since access tokens have an expiry and need to be refreshed.  We decided at the meeting to tie the session key's expiration to the access token's expiration and refresh the session key at the same time.

# Preferences Edge Proxy Refinement I

This describes the requests, responses, and payloads for the workflow where a
static site makes save/retrieve requests of a secured preferences server.  For
example, the static site may have incorporated UIO+ into their pages.  The user

> wants to save or retrieve their UIO+ preferences.

Reviewer: I believe you mean to say UIO rather than UIO+. UIO+ is the chrome extension, which at some point would also make sense to connect to the prefs server.

Author: Right.  Done.

> The outline of the workflow is taken verbatim from the Fluid wiki page titled

Reviewer: I think it would make sense to recontectualize the workflow here, as portions that talk about GPII make it a bit confusing. For example in point 2. it says "implemented by the GPII authorization server" which is something that will no longer be happening, and seems like it will be done via Google or other SSO provider.
